test/docker/compose.go

```diff
@@ -513,3 +485,112 @@ func (d *Compose) Start(ctx context.Context) (*DockerCompose, error) {
 
 	return &DockerCompose{network, containers}, nil
 }
+
+func (d *Compose) With1NodeCluster() *Compose {
+	d.withWeaviateCluster = true
+	d.size = 1
+	return d
+}
+
+func (d *Compose) With2NodeCluster() *Compose {
+	d.withWeaviateCluster = true
+	d.size = 2
+	return d
+}
+
+func (d *Compose) With3NodeCluster() *Compose {
+	d.withWeaviateCluster = true
+	d.size = 3
+	return d
+}
+
+func (d *Compose) startCluster(ctx context.Context, size int, settings map[string]string) ([]*DockerContainer, error) {
+	if size == 0 || size > 3 {
+		return nil, nil
+	}
+
+	raft_join := "node1,node2,node3"
+	if size == 1 {
+		raft_join = "node1"
+	} else if size == 2 {
+		raft_join = "node1,node2"
+	}
+
+	cs := make([]*DockerContainer, size)
+	image := os.Getenv(envTestWeaviateImage)
+	networkName := settings["network"]
+	if d.withWeaviateBasicAuth {
+		settings["CLUSTER_BASIC_AUTH_USERNAME"] = d.withWeaviateBasicAuthUsername
+		settings["CLUSTER_BASIC_AUTH_PASSWORD"] = d.withWeaviateBasicAuthPassword
+	}
+	if d.withWeaviateAuth {
+		settings["AUTHENTICATION_OIDC_ENABLED"] = "true"
+		settings["AUTHENTICATION_OIDC_CLIENT_ID"] = "wcs"
+		settings["AUTHENTICATION_OIDC_ISSUER"] = "https://auth.wcs.api.semi.technology/auth/realms/SeMI"
+		settings["AUTHENTICATION_OIDC_USERNAME_CLAIM"] = "email"
+		settings["AUTHENTICATION_OIDC_GROUPS_CLAIM"] = "groups"
+		settings["AUTHORIZATION_ADMINLIST_ENABLED"] = "true"
+		settings["AUTHORIZATION_ADMINLIST_USERS"] = "ms_2d0e007e7136de11d5f29fce7a53dae219a51458@existiert.net"
+	}
+	settings["RAFT_PORT"] = "8300"
+	settings["RAFT_INTERNAL_RPC_PORT"] = "8301"
+	settings["RAFT_JOIN"] = raft_join
+	settings["RAFT_BOOTSTRAP_EXPECT"] = strconv.Itoa(d.size)
+
+	// first node
+	config1 := copySettings(settings)
+	config1["CLUSTER_HOSTNAME"] = "node1"
+	config1["CLUSTER_GOSSIP_BIND_PORT"] = "7100"
+	config1["CLUSTER_DATA_BIND_PORT"] = "7101"
+	eg := errgroup.Group{}
+	eg.Go(func() (err error) {
+		if cs[0], err = startWeaviate(ctx, d.enableModules, d.defaultVectorizerModule,
+			config1, networkName, image, Weaviate1, d.withWeaviateExposeGRPCPort); err != nil {
+			return errors.Wrapf(err, "start %s", Weaviate1)
+		}
+		return nil
+	})
+
+	if size > 1 {
+		config2 := copySettings(settings)
+		config2["CLUSTER_HOSTNAME"] = "node2"
+		config2["CLUSTER_GOSSIP_BIND_PORT"] = "7102"
+		config2["CLUSTER_DATA_BIND_PORT"] = "7103"
+		config2["CLUSTER_JOIN"] = fmt.Sprintf("%s:7100", Weaviate1)
+		eg.Go(func() (err error) {
+			time.Sleep(time.Second * 3)
```

Comment:
you don't need to add sleep here to wait for the container to be ready, this is done by testcontainers using wait strategies

Reply:
The container depends on node1 to start member list initialization. if we don't introduce a delay, the container will just skip because unlike previoius approach the containers are started concurrently so that we don't control the order or initialization. In addition the definiton of ready has changed in the raft setup because node depend on each other